### PR TITLE
Split fv3fit.ModelTrainingConfig into new objects for data and training config

### PR DESCRIPTION
**Pull request by mcgibbon**
Currently ModelTrainingConfig is responsible for holding configuration options for training, training data loading, and validation data loading. This PR splits those responsibilities into two separate classes, and three separate instances (training and validation data configuration use the same class).

In this PR this has been implemented to be backwards-compatible.

Added public API:
- Added classes `TrainingConfig`, `SklearnTrainingConfig`, `KerasTrainingConfig`, and `DataConfig` to `fv3fit`

Refactored public API:
- Bulleted list of removed or refactored symbols, such as changes to name, type, behavior, argument, etc. Be cautious about doing these and discuss with team more broadly.

Significant internal changes:
- Added `register_keras_trainer` decorator for registering new keras training classes
- load_data_sequence now takes in only one `DataConfig` argument, instead of a data path and ModelTrainingConfig
- a routine legacy_config_to_data_config is added to be used temporarily by test code to convert ModelTrainingConfig objects into DataConfig objects. The tests using this should be later refactored to create TrainingConfig and DataConfig objects directly.
- Moved tightly coupled code involving validation data into `_shared/config.py`. This code is quite prone to error, only being enabled for Keras training. It should be deleted in a follow-on PR, instead using a `DataConfig` for validation data in a way that works the same for Keras and Sklearn training.

- [x] Tests added

You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

**Comment by nbren12** [comment]
This is looking real nice @mcgibbon! Would it be worth grouping the objects into a top-level configuration like the prognostic run does: https://github.com/VulcanClimateModeling/fv3net/blob/a24d15ef763a2a157e62738dbf09738cb85fe8ad/workflows/prognostic_c48_run/runtime/config.py#L21.

It's a nice way to collect the interface and centralize the documentation.

**Comment by mcgibbon** [comment]
> This is looking real nice @mcgibbon! Would it be worth grouping the objects into a top-level configuration like the prognostic run does

Yeah, that looks pretty nice!

I'll have to see depending on the backwards-compatibility-breaking change we go for. If we want to and can use a central object to contain all the configs in one yaml file, that is probably the way to go. It looks nice for external API presentation. The data configs basically immediately get transformed into batches, and then only the training config gets persisted, so there's no reason/need currently to group the configs internally in the application.

This will also depend on some things like, I want to refactor the configuration saving so that the configuration is attached to the model rather than being logged. Right now we duplicate "input variables" info between the model and the train_config.yaml, which makes it confusing when using the offline diagnostics on their own (in my experience). This may mean we want to provide the DataConfig to the training just so the model can save it, it depends on what people think is important.

I'll leave this potential change for now until those needs are more clear.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/_shared/config.py:327]: Leftover print

File: external/fv3fit/fv3fit/train.py
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/train.py:106]:
```suggestion
    train_config.dump(os.path.join(args.output_data_path, "train.yml"))
```

File: external/fv3fit/fv3fit/_shared/config.py
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/_shared/config.py:61]: `timesteps_source` and `validation_timesteps` deleted

File: external/fv3fit/fv3fit/_shared/config.py
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/_shared/config.py:94]: Remove extra args that are no longer in this dataclass

**Review by AnnaKwa — CHANGES_REQUESTED** [review]
Thanks, the splitting up of the big config looks good! Just one main request regarding the dump/load we talked about.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/config.py:119]: I think it could mislead someone who sees this in the Keras config into thinking keras training supports this, so I'd like to leave it out until we make the support uniform.

**Review by AnnaKwa — APPROVED** [review]
Actually, I'm pretty sure I don't need to rereview for the requested change, so I'm approving with that request about removing dump/load. Could you run the integration test before merging just to make sure no breaking changes were missed?

File: external/fv3fit/fv3fit/_shared/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/config.py]: Done. I refactored the tests to ones that ensure dataclasses.asdict, cls(**as_dict), and yaml.safe_dump all work on the config classes.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/config.py]: Removed.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/config.py:77]: Thanks, done. Added docstrings to the subclasses which include their additional options.

File: external/fv3fit/fv3fit/_shared/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/config.py]: Removed, thanks!

File: external/fv3fit/fv3fit/_shared/config.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/_shared/config.py]: Removed.

File: external/fv3fit/fv3fit/train.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/train.py:96]: I'm not sure what you mean, this is a configuration object, not a model object?

File: external/fv3fit/fv3fit/train.py
Comment (mcgibbon) [reply external/fv3fit/fv3fit/train.py]: I changed these new ones to use `.yaml` since that's what's [officially recommended](https://yaml.org/faq.html).

File: workflows/offline_ml_diags/tests/test_compute_diags.py
Comment (mcgibbon) [reply workflows/offline_ml_diags/tests/test_compute_diags.py:14]: Yeah I did it just as a way to tell myself "this is bad, fix/delete it", I'll make this more explicit by adding TODO comments just above these imports. My next task is switching this out for the new config classes.

**Pull request merged**
